packages/kit/src/core/env.js

```diff
@@ -25,46 +34,55 @@ export function create_static_module(id, env) {
 }
 
 /**
- * @param {'public' | 'private'} type
+ * @param {EnvType} type
  * @param {Record<string, string> | undefined} dev_values If in a development mode, values to pre-populate the module with.
  */
 export function create_dynamic_module(type, dev_values) {
 	if (dev_values) {
-		const objectKeys = Object.entries(dev_values).map(
+		const keys = Object.entries(dev_values).map(
 			([k, v]) => `${JSON.stringify(k)}: ${JSON.stringify(v)}`
 		);
-		return `const env = {\n${objectKeys.join(',\n')}\n}\n\nexport { env }`;
+		return `export const env = {\n${keys.join(',\n')}\n}`;
 	}
 	return `export { env } from '${runtime_base}/env-${type}.js';`;
 }
 
 /**
- * @param {string} id
- * @param {Record<string, string>} env
+ * @param {EnvType} id
+ * @param {EnvData} env
  * @returns {string}
  */
 export function create_static_types(id, env) {
-	const declarations = Object.keys(env)
+	const declarations = Object.keys(env[id])
 		.filter((k) => valid_identifier.test(k))
 		.map((k) => `\texport const ${k}: string;`)
 		.join('\n');
 
-	return `declare module '${id}' {\n${declarations}\n}`;
+	return `declare module '$env/static/${id}' {\n${declarations}\n}`;
 }
 
 /**
- * @param {string} id
- * @param {Record<string, string>} env
+ * @param {EnvType} id
+ * @param {EnvData} env
  * @returns {string}
  */
 export function create_dynamic_types(id, env) {
-	const properties = Object.keys(env)
+	const properties = Object.keys(env[id])
 		.filter((k) => valid_identifier.test(k))
 		.map((k) => `\t\t${k}: string;`);
 
-	properties.push(`\t\t[key: string]: string | undefined;`);
+	properties.push(
+		`\t\t[key: \`${env.prefix}\${string}\`]: ${
+			id === 'public' ? 'string | undefined' : 'undefined'
+		};`
+	);
+
+	if (id === 'private') {
+		properties.push(`\t\t[key: string]: string | undefined;`);
+	}
```

Comment:
I think it might make more sense to combine this into

```js
if (id === 'private') {
  // push [key: string] signature
} else {
  // push [`PUBLICPREFIX_${string}`: string] signature
}
```
(whats the minimum required TS version for this btw? I think it's in our range, right?)

Reply:
I can see that, but it'll look more like this

```js
if (id === 'private') {
  // push { [`PUBLICPREFIX_${string}`: string]: undefined } signature
  // push { [key: string]: string | undefined } signature
} else {
  // push { [`PUBLICPREFIX_${string}`: string]: string | undefined } signature
}
```

I believe it is within our range, minimum version for this was 4.4 where [template literal in index signature](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#symbol-template-signatures) was added

Reply:
Mhm for some reason I thought that private dynamic also includes the public variables for convenience - nevermind then.